### PR TITLE
Add wsfc manger and agent

### PR DESCRIPTION
**Pull request by ning-yang**
wsfc(window failover cluster) agent will listen to healthcheck ping and
based on the ip from the playload, it will check whether the ip exists.
If it find it, it will responce healthy (as "1").

### CODE REVIEW

File: GCEWindowsAgent/wsfc.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+type wsfcManager struct {
+	agentNewState agentState
+	agentNewPort  string
+	agent         healthAgent
+}
+
+// Create new wsfcManager based on metadata
+func newWsfcManager(newMetadata *metadataJSON) *wsfcManager {
+	newState := stopped
+	if newMetadata.Instance.Attributes.EnableWSFC ||
```
Comment (adjackura) [inline GCEWindowsAgent/wsfc.go]: what if EnableWSFC is false but there are WSFCAddresses?

File: GCEWindowsAgent/wsfc.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+	return &wsfcManager{agentNewState: newState, agentNewPort: newPort, agent: getWsfcAgentInstance()}
+}
+
+// Implement manager.diff()
+func (m *wsfcManager) diff() bool {
+	return m.agentNewState != m.agent.getState() || m.agentNewPort != m.agent.getPort()
+}
+
+// Implement manager.disabled().
+// wsfc manager is always enabled.
```
Comment (adjackura) [inline GCEWindowsAgent/wsfc.go]: add a note to why this is (you have a long running process)

File: GCEWindowsAgent/wsfc.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+// Stop agent. Will wait for all existing request to be completed.
+func (a *wsfcAgent) stop() error {
+	if a.state == stopped {
+		logger.Info("wsfc agent already stopped.")
+		return nil
+	}
+
+	logger.Info("wsfc agent stopped.")
```
Comment (adjackura) [inline GCEWindowsAgent/wsfc.go]: logger.Info("Stopping wsfc agent...")

File: GCEWindowsAgent/wsfc.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+// Start agent and taking tcp request
+func (a *wsfcAgent) run(errc chan error) {
+	if a.state == running {
+		logger.Infoln("wsfc agent is already running")
+		errc <- nil
+		return
+	}
+
+	logger.Info("wsfc agent starting")
```
Comment (adjackura) [inline GCEWindowsAgent/wsfc.go]: logger.Info("Starting wsfc agent...")

File: GCEWindowsAgent/wsfc.go (hunk 1 above)
Comment (ning-yang) [inline GCEWindowsAgent/wsfc.go]: WSFCAddresses always takes priority as an advanced setting. Same for the address manager will always respect this setting and ignore enable-wsfc if it is set.

File: GCEWindowsAgent/wsfc.go (hunk 4 above)
Comment (ning-yang) [inline GCEWindowsAgent/wsfc.go]: done

File: GCEWindowsAgent/wsfc.go (hunk 3 above)
Comment (ning-yang) [inline GCEWindowsAgent/wsfc.go]: done

**Comment by ning-yang** [comment]
Agent Run logic has been updated.

File: GCEWindowsAgent/wsfc.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+		}
+
+		return m.startAgent()
+	}
+
+	return nil
+}
+
+// Start health agent in goroutine
+func (m *wsfcManager) startAgent() error {
```
Comment (ning-yang) [inline GCEWindowsAgent/wsfc.go]: originally, there is some additional logic to create channel and pass it to agent run/stop. But we no longer have it now. removed

File: GCEWindowsAgent/wsfc.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+func (a *wsfcAgent) handleHealthCheckRequest(conn net.Conn) {
+	defer conn.Close()
+	defer a.waitGroup.Done()
+	conn.SetDeadline(time.Now().Add(time.Second))
+
+	buf := make([]byte, 1024)
+	// Read the incoming connection into the buffer.
+	reqLen, err := conn.Read(buf)
+	if err != nil {
+		logger.Errorln("error on processing request:", err)
```
Comment (ning-yang) [inline GCEWindowsAgent/wsfc.go]: done

**Pull request merged**